File: app/src/main/java/org/tasks/Notifier.java
package org.tasks;

import static android.content.Intent.FLAG_ACTIVITY_MULTIPLE_TASK;
import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static com.google.common.collect.Lists.transform;
import static org.tasks.time.DateTimeUtils.currentTimeMillis;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.support.v4.app.NotificationCompat;
import com.todoroo.andlib.utility.AndroidUtilities;
import com.todoroo.astrid.activity.MainActivity;
import com.todoroo.astrid.api.Filter;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import com.todoroo.astrid.reminders.ReminderService;
import com.todoroo.astrid.voice.VoiceOutputAssistant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.inject.Inject;
import org.tasks.injection.ForApplication;
import org.tasks.jobs.NotificationQueueEntry;
import org.tasks.notifications.AudioManager;
import org.tasks.notifications.NotificationManager;
import org.tasks.notifications.TelephonyManager;
import org.tasks.preferences.Preferences;
import org.tasks.ui.CheckBoxes;
import timber.log.Timber;

public class Notifier {

  private final Context context;
  private final TaskDao taskDao;
  private final NotificationManager notificationManager;
  private final TelephonyManager telephonyManager;
  private final AudioManager audioManager;
  private final VoiceOutputAssistant voiceOutputAssistant;
  private final Preferences preferences;
  private final CheckBoxes checkBoxes;

  @Inject
  public Notifier(
      @ForApplication Context context,
      TaskDao taskDao,
      NotificationManager notificationManager,
      TelephonyManager telephonyManager,
      AudioManager audioManager,
      VoiceOutputAssistant voiceOutputAssistant,
      Preferences preferences,
      CheckBoxes checkBoxes) {
    this.context = context;
    this.taskDao = taskDao;
    this.notificationManager = notificationManager;
    this.telephonyManager = telephonyManager;
    this.audioManager = audioManager;
    this.voiceOutputAssistant = voiceOutputAssistant;
    this.preferences = preferences;
    this.checkBoxes = checkBoxes;
  }

  public void triggerFilterNotification(final Filter filter) {
    List<Task> tasks = taskDao.fetchFiltered(filter);
    int count = tasks.size();
    if (count == 0) {
      return;
    }

    Intent intent = new Intent(context, MainActivity.class);
    intent.setFlags(FLAG_ACTIVITY_NEW_TASK | FLAG_ACTIVITY_MULTIPLE_TASK);
    intent.putExtra(MainActivity.OPEN_FILTER, filter);
    PendingIntent pendingIntent =
        PendingIntent.getActivity(
            context, filter.listingTitle.hashCode(), intent, PendingIntent.FLAG_UPDATE_CURRENT);

    String summaryTitle =
        context.getResources().getQuantityString(R.plurals.task_count, count, count);
    NotificationCompat.InboxStyle style =
        new NotificationCompat.InboxStyle().setBigContentTitle(summaryTitle);
    int maxPriority = 3;
    for (Task task : tasks) {
      style.addLine(task.getTitle());
      maxPriority = Math.min(maxPriority, task.getPriority());
    }

    NotificationCompat.Builder builder =
        new NotificationCompat.Builder(context, NotificationManager.NOTIFICATION_CHANNEL_TASKER)
            .setSmallIcon(R.drawable.ic_done_all_white_24dp)
            .setCategory(NotificationCompat.CATEGORY_REMINDER)
            .setTicker(summaryTitle)
            .setContentTitle(summaryTitle)
            .setContentText(filter.listingTitle)
            .setContentIntent(pendingIntent)
            .setAutoCancel(true)
            .setWhen(currentTimeMillis())
            .setShowWhen(true)
            .setColor(checkBoxes.getPriorityColor(maxPriority))
            .setGroupSummary(true)
            .setGroup(filter.listingTitle)
            .setStyle(style);

    notificationManager.notify(filter.listingTitle.hashCode(), builder, true, false, false);
  }

  public void triggerTaskNotification(long id, int type) {
    org.tasks.notifications.Notification notification = new org.tasks.notifications.Notification();
    notification.taskId = id;
    notification.type = type;
    notification.timestamp = currentTimeMillis();
    triggerNotifications(Collections.singletonList(notification), true);
  }

  public void triggerTaskNotifications(List<? extends NotificationQueueEntry> entries) {
    triggerNotifications(transform(entries, NotificationQueueEntry::toNotification), true);
  }

  private void triggerNotifications(
      List<org.tasks.notifications.Notification> entries, boolean alert) {
    List<org.tasks.notifications.Notification> notifications = new ArrayList<>();
    boolean ringFiveTimes = false;
    boolean ringNonstop = false;
    for (int i = 0; i < entries.size(); i++) {
      org.tasks.notifications.Notification entry = entries.get(i);
      Task task = taskDao.fetch(entry.taskId);
      if (task == null) {
        continue;
      }
      if (entry.type != ReminderService.TYPE_RANDOM) {
        ringFiveTimes |= task.isNotifyModeFive();
        ringNonstop |= task.isNotifyModeNonstop();
      }
      NotificationCompat.Builder notification = notificationManager.getTaskNotification(entry);
      if (notification != null) {
        notifications.add(entry);
      }
    }

    if (notifications.isEmpty()) {
      return;
    } else {
      Timber.d("Triggering %s", notifications);
    }

    notificationManager.notifyTasks(notifications, alert, ringNonstop, ringFiveTimes);

    if (alert
        && preferences.getBoolean(R.string.p_voiceRemindersEnabled, false)
        && !ringNonstop
        && !audioManager.notificationsMuted()
        && telephonyManager.callStateIdle()) {
      for (org.tasks.notifications.Notification notification : notifications) {
        AndroidUtilities.sleepDeep(2000);
        voiceOutputAssistant.speak(
            notificationManager.getTaskNotification(notification).build().tickerText.toString());
      }
    }
  }
}

File: app/src/main/java/org/tasks/Notifier.kt
package org.tasks

import android.app.PendingIntent
import android.content.Context
import androidx.core.app.NotificationCompat
import com.todoroo.astrid.api.Filter
import com.todoroo.astrid.voice.VoiceOutputAssistant
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.delay
import org.tasks.data.Alarm
import org.tasks.data.Alarm.Companion.TYPE_GEO_ENTER
import org.tasks.data.Alarm.Companion.TYPE_GEO_EXIT
import org.tasks.data.Geofence
import org.tasks.data.TaskDao
import org.tasks.intents.TaskIntents
import org.tasks.notifications.AudioManager
import org.tasks.notifications.Notification
import org.tasks.notifications.NotificationManager
import org.tasks.notifications.TelephonyManager
import org.tasks.preferences.Preferences
import org.tasks.themes.ColorProvider
import org.tasks.time.DateTimeUtils
import timber.log.Timber
import javax.inject.Inject
import kotlin.math.min

class Notifier @Inject constructor(
        @param:ApplicationContext private val context: Context,
        private val taskDao: TaskDao,
        private val notificationManager: NotificationManager,
        private val telephonyManager: TelephonyManager,
        private val audioManager: AudioManager,
        private val voiceOutputAssistant: VoiceOutputAssistant,
        private val preferences: Preferences) {

    private val colorProvider: ColorProvider = ColorProvider(context, preferences)

    suspend fun triggerFilterNotification(filter: Filter) {
        val tasks = taskDao.fetchFiltered(filter)
        val count = tasks.size
        if (count == 0) {
            return
        }
        val intent = TaskIntents.getTaskListIntent(context, filter)
        val pendingIntent = PendingIntent.getActivity(
            context,
            filter.title.hashCode(),
            intent,
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
        val summaryTitle = context.resources.getQuantityString(R.plurals.task_count, count, count)
        val style = NotificationCompat.InboxStyle().setBigContentTitle(summaryTitle)
        var maxPriority = 3
        for (task in tasks) {
            style.addLine(task.title)
            maxPriority = min(maxPriority, task.priority)
        }
        val builder = NotificationCompat.Builder(context, NotificationManager.NOTIFICATION_CHANNEL_TASKER)
                .setSmallIcon(R.drawable.ic_done_all_white_24dp)
                .setCategory(NotificationCompat.CATEGORY_REMINDER)
                .setTicker(summaryTitle)
                .setContentTitle(summaryTitle)
                .setContentText(filter.title)
                .setContentIntent(pendingIntent)
                .setAutoCancel(true)
                .setWhen(DateTimeUtils.currentTimeMillis())
                .setShowWhen(true)
                .setColor(colorProvider.getPriorityColor(maxPriority, true))
                .setGroupSummary(true)
                .setGroup(filter.title)
                .setStyle(style)
        notificationManager.notify(filter.title.hashCode().toLong(), builder, alert = true, nonstop = false, fiveTimes = false)
    }

    suspend fun triggerNotifications(place: Long, geofences: List<Geofence>, arrival: Boolean) =
            geofences
                    .filter { if (arrival) it.isArrival else it.isDeparture }
                    .map {
                        Notification().apply {
                            taskId = it.task
                            type = if (arrival) TYPE_GEO_ENTER else TYPE_GEO_EXIT
                            timestamp = DateTimeUtils.currentTimeMillis()
                            location = place
                        }
                    }
                    .let { triggerNotifications(it) }

    suspend fun triggerNotifications(entries: List<Notification>) {
        var ringFiveTimes = false
        var ringNonstop = false
        val notifications = entries
                .filter {
                    taskDao.fetch(it.taskId)
                            ?.let { task ->
                                if (it.type != Alarm.TYPE_RANDOM) {
                                    ringFiveTimes = ringFiveTimes or task.isNotifyModeFive
                                    ringNonstop = ringNonstop or task.isNotifyModeNonstop
                                }
                                notificationManager.getTaskNotification(it) != null
                            }
                            ?: false
                }
                .takeLast(NotificationManager.MAX_NOTIFICATIONS)

        if (notifications.isEmpty()) {
            return
        }

        Timber.d("Triggering $notifications")

        notificationManager.notifyTasks(notifications, true, ringNonstop, ringFiveTimes)

        if (preferences.getBoolean(R.string.p_voiceRemindersEnabled, false)
                && !ringNonstop
                && !audioManager.notificationsMuted()
                && telephonyManager.callStateIdle()) {
            notifications
                    .mapNotNull {
                        notificationManager.getTaskNotification(it)?.build()?.tickerText?.toString()
                    }
                    .forEach {
                        delay(2000)
                        voiceOutputAssistant.speak(it)
                    }
        }
    }
}